### PR TITLE
[DTensor] support split op on Partial placement

### PR DESCRIPTION
**Pull request by XilunWu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #158112
* #158051
* __->__ #157991

**Summary**
To enable use case where the input DTensor to `split` op has `Partial()` placement,
this PR treats `Partial()` in the same way with `Replicate()`. That means, `split` op
only unshards the `Shard(dim=x)` if `x == split_dim` and keep other placement
untouched.

**Test**
Added a new test because `test_dtensor_ops` doesn't test `Partial()` placement.
`pytest test/distributed/tensor/test_tensor_ops.py -s -k test_split_on_partial`

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @tianyu-l

### CODE REVIEW

File: test/distributed/tensor/test_tensor_ops.py [hunk 1]
```diff
@@ -708,6 +709,46 @@ def test_slice(self):
...
+        replicate_dt_full_tensor_list = [dt.full_tensor() for dt in partial_dt_list]
+        try:
+            for replicate_tensor, replicate_dt_full_tensor in zip(
+                replicate_tensor_list, replicate_dt_full_tensor_list
+            ):
+                self.assertEqual(replicate_tensor, replicate_dt_full_tensor)
+        except AssertionError:
+            print(
+                f"failing case reduce_op={reduce_op} split_size={split_size} split_dim={split_dim}"
+            )
```
Comment (ezyang) [inline test/distributed/tensor/test_tensor_ops.py]: I don't understand, what's the point of this test case then?

Comment (ezyang) [reply test/distributed/tensor/test_tensor_ops.py]: The other contents seem fine though

File: test/distributed/tensor/test_tensor_ops.py [hunk 2]
```diff
@@ -708,6 +709,46 @@ def test_slice(self):
...
+            self._test_split_on_partial,
+        )
+
+    def _test_split_on_partial(self, reduce_op: str, split_size: int, split_dim: int):
+        torch.manual_seed(self.rank)
+        mesh = init_device_mesh(self.device_type, (self.world_size,))
+
+        partial_tensor = torch.randn(8, 8, device=self.device_type)
+        replicate_tensor = partial_tensor.detach().clone()
+        replicate_tensor = funcol.all_reduce(
```
Comment (ezyang) [inline test/distributed/tensor/test_tensor_ops.py:729]: A more generic testing thought: it seems to me that this sort of "do the operation with dtensor but also do it with plain tensor" feels like it would be pretty common, can we not automate it?  For example, we don't have to manually setup the replicate_tensor and collective: instead, we can simply do partial_dt.full_tensor().split(split_size, dim=split_dim). We're simply testing that we check this diagram commutes:

```
             split()
  partial_dt ---------> B
      |                  |
      |                  | full_tensor()
full_tensor()            |
      |                  |
      v                  v
      C ---------> result
             split()
```

File: test/distributed/tensor/test_tensor_ops.py (hunk 1 above)
Comment (XilunWu) [reply test/distributed/tensor/test_tensor_ops.py]: the try-catch should've been removed from the PR. VSCode buffer wasn't correctly saved to file syetem before ghstack-ing. :-(

File: test/distributed/tensor/test_tensor_ops.py (hunk 2 above)
Comment (XilunWu) [reply test/distributed/tensor/test_tensor_ops.py:729]: Good point, will add a base PR to introduce such test util. IIRC we have similar logics in each separate test file but haven't merged them into one shared util.

File: test/distributed/tensor/test_tensor_ops.py [hunk 3]
```diff
@@ -708,6 +710,41 @@ def test_slice(self):
...
+                "reduce_op": ["sum", "avg", "product", "min", "max"],
+                "split_size": [2, 3, 4],
+                "split_dim": [0, 1],
+            },
+            self._test_split_on_partial,
+        )
+
+    def _test_split_on_partial(self, reduce_op: str, split_size: int, split_dim: int):
+        torch.manual_seed(self.rank)
+        mesh = init_device_mesh(self.device_type, (self.world_size,))
```
Comment (zpcore) [inline test/distributed/tensor/test_tensor_ops.py:725]: Can we try on 2d mesh just to be safe with the allreduce?

File: test/distributed/tensor/test_tensor_ops.py (hunk 2 above)
Comment (zpcore) [reply test/distributed/tensor/test_tensor_ops.py:729]: I prefer this way to do test:
```
a = tensor(...)
a_dt = distribute_tensor(a, mesh, placement)
b = op(a)
b_dt = op(a_dt)
assert b==b_dt.full_tensor()
```
Feel like this is more straight forward.  However, `distribute_tensor` doesn't allow Partial() currently.

Comment (XilunWu) [reply test/distributed/tensor/test_tensor_ops.py:729]: @zpcore the reason we don't allow `Partial` in `distribute_tensor` is that `Partial` is not supposed to be used directly by users and it's only for DTensor's internal. To construct a Partial DTensor, we can use `DTensor.from_local()`.

**Review by zpcore — APPROVED** [review]
LGTM!

File: test/distributed/tensor/test_tensor_ops.py (hunk 2 above)
Comment (XilunWu) [reply test/distributed/tensor/test_tensor_ops.py:729]: implemented @ezyang's suggestion in https://github.com/pytorch/pytorch/pull/158112
